packages/trader/src/exchange-adapters/UniswapV3Trader.ts

```diff
@@ -0,0 +1,59 @@
+import winston from "winston";
+import Web3 from "web3";
+import BigNumber from "bignumber.js";
+
+const bn = require("bignumber.js"); // Big number that comes with web3 does not support square root.
+bn.config({ EXPONENTIAL_AT: 999999, DECIMAL_PLACES: 40 });
+
+const { encodePriceSqrt } = require("@uma/common");
+const ExchangeAdapterInterface = require("./ExchangeAdapterInterface");
+const { getTruffleContract } = require("@uma/core");
+
+class UniswapV3Trader implements InstanceType<typeof ExchangeAdapterInterface> {
+  readonly tradeDeadline: number;
+  readonly UniswapV2Broker: any;
+  uniswapPair: any;
+
+  constructor(
+    readonly logger: winston.Logger,
+    readonly web3: Web3,
+    readonly uniswapPoolAddress: string,
+    readonly uniswapRouterAddress: string,
+    readonly dsProxyManager: any
+  ) {
+    this.logger = logger;
+    this.web3 = web3;
+    this.uniswapPoolAddress = uniswapPoolAddress;
+    this.uniswapRouterAddress = uniswapRouterAddress;
+
+    this.dsProxyManager = dsProxyManager;
+
+    this.tradeDeadline = 10 * 60 * 60;
+
+    this.UniswapV2Broker = getTruffleContract("UniswapV3Broker", this.web3);
+  }
+  async tradeMarketToDesiredPrice(desiredPrice: BigNumber) {
+    const callCode = this.UniswapV2Broker.bytecode;
+
+    const contract = new this.web3.eth.Contract(this.UniswapV2Broker.abi);
+
+    const callData = contract.methods
+      .swapToPrice(
+        false, // tradingAsEOA. Set as false as this is executed as a DSProxy.
+        this.uniswapPoolAddress, // address of the pool to uniswap v3 trade against.
+        this.uniswapRouterAddress, // address of the uniswap v3 router to route the trade.
+        encodePriceSqrt(desiredPrice, this.web3.utils.toWei("1")), // sqrtRatioTargetX96 target, encoded price.
+        this.dsProxyManager.getDSProxyAddress(), // to: the output of the trade will send the tokens to the DSProxy.
+        Number((await this.web3.eth.getBlock("latest")).timestamp) + this.tradeDeadline // Deadline in the future
+      )
+      .encodeABI();
+
+    try {
+      return await this.dsProxyManager.callFunctionOnNewlyDeployedLibrary(callCode, callData);
+    } catch (error) {
+      return error;
+    }
```

Comment:
Does this satisfy the declared return type of the interface (`void`)? What types are you returning here? Also, just out of curiosity, why are we returning the error rather than just throwing?

Reply:
you right, it does not. This method will either return what `callFunctionOnNewlyDeployedLibrary` returns (which contains transaction information, tx config and return values from function calls on contract) OR returns an error. within the `RangeTrader` the return type is checked if it is an `instanceOfError`. if it is then error logic is handled in that layer.

An alternative pattern would be to just throw (or not even wrap in try catch here) and then put the try-catch logic one layer up within the `RangeTrader`. the advantage of this is that we dont need these `try-catch` blocks within the exchange-adapters, thereby placing the burden on the caller of this adapter. WDYT?

Reply:
I've updated the type in the interface to return `Promise<Error | { transactionHash: string; any: any }>`. In this way, the `RangeTrader` can deal with the error case by checking the type returned. the `RangeTrader` uses the `transactionHash` in logging. I added the `any:any` to enable us to pass extra props in the return object without any issues (additional props are passed back in the `runTransaction` Util PR, for example).

Reply:
That sounds great! How are the `any: any` parts used? Is it just for logging? What does the caller do with the `any` field? If it's just for logging, could we have a type like the following so you require a transaction hash and also allow any other string -> string fields.
```ts
Promise<Error | { transactionHash: string; [key: string]: string; }>
```

If it really could be anything, we can use `any`, but I think it's probably preferable to try and describe what shape the extra props are going to come in so they can be more effectively used by the calling code.

Reply:
I'm not exactly sure what the structure will be after the refactor in the `runTransaction` PR. this is why I left it as `any:any`. I think it could be both an `Error` or a string-keyed object describing transaction outputs.